server.c

```diff
@@ -978,24 +987,59 @@ server_queuereader(void *d)
 		}
 
 		for (; *metric != NULL; metric++) {
-			len = *(size_t *)(*metric);
-			/* Write to the stream, this may not succeed completely due
-			 * to flow control and whatnot, which the docs suggest need
-			 * resuming to complete.  So, use a loop, but to avoid
-			 * getting endlessly stuck on this, only try a limited
-			 * number of times for a single metric. */
-			for (cnt = 0, p = *metric + sizeof(size_t); cnt < 10; cnt++) {
-				if ((slen = self->strm->strmwrite(self->strm, p, len)) != len) {
-					if (slen >= 0) {
-						p += slen;
-						len -= slen;
-					} else if (errno != EINTR) {
+			size_t metricscount;
+			if (buf == NULL || *(size_t *)(*metric) >= self->qreadersendsize) {
+				len = *(size_t *)(*metric);
+				metricscount = 1;
+				p = *metric + sizeof(size_t);
+				/* Write to the stream, this may not succeed completely due
+				 * to flow control and whatnot, which the docs suggest need
+				 * resuming to complete.  So, use a loop, but to avoid
+				 * getting endlessly stuck on this, only try a limited
+				 * number of times for a single metric. */
+				for (cnt = 0; cnt < 10; cnt++) {
+					if ((slen = self->strm->strmwrite(self->strm, p, len)) != len) {
+						if (slen >= 0) {
+							p += slen;
+							len -= slen;
+						} else if (errno != EINTR) {
+							break;
+						}
+						/* allow the remote to catch up */
+						usleep((50 + (rand() % 150)) * 1000);  /* 50ms - 200ms */
+					} else {
+						break;
+					}
+				}
+			} else { /* batch send */
+				metricscount = 0;
+				const char **pmetric;
+				len = 0;
+				/* fill metrics buffer for send */
+				for (pmetric = metric; *pmetric != NULL; pmetric++) {
+					if (len + *(size_t *)(*pmetric) >= self->qreadersendsize) {
+						break;
+					}
+					metricscount++;
+					memcpy(buf + len, *pmetric + sizeof(size_t), *(size_t *) (*pmetric));
```

Comment:
May be, not testing buffered writer - can't detect errors transparently.
Reason for this - copy some data to temporary buffer in userspace and send in one write call cheaper than many write calls without additional copy. You can see this in profile callgraphs.

Reply:
for not to be only applying to the single case, if buffer has a sufficient size to accommodate the metric, but not too large to avoid fragmentation when sending. 32 kilobytes is enough.